vernac/egramcoq.ml

```diff
@@ -231,25 +242,59 @@ type (_, _) entry =
 | TTName : ('self, Misctypes.lname) entry
 | TTReference : ('self, reference) entry
 | TTBigint : ('self, Constrexpr.raw_natural_number) entry
-| TTConstr : prod_info * 'r target -> ('r, 'r) entry
+| TTConstr : notation_entry * prod_info * 'r target -> ('r, 'r) entry
 | TTConstrList : prod_info * Tok.t list * 'r target -> ('r, 'r list) entry
 | TTPattern : int -> ('self, cases_pattern_expr) entry
 | TTOpenBinderList : ('self, local_binder_expr list) entry
 | TTClosedBinderList : Tok.t list -> ('self, local_binder_expr list list) entry
 
 type _ any_entry = TTAny : ('s, 'r) entry -> 's any_entry
 
+let register_custom_grammar s =
+  let sc = "constr:"^s in
+  let sp = "pattern:"^s in
+  let wit1 = Genarg.create_arg sc in
+  let wit2 = Genarg.create_arg sp in
+  let c = (Gram.entry_create sc : constr_expr Gram.entry) in
+  let p = (Gram.entry_create sp : cases_pattern_expr Gram.entry) in
+  register_grammar wit1 c;
+  register_grammar wit2 p;
+  register_grammars_by_name s [AnyEntry c; AnyEntry p];
+  (c,p)
+
+let find_custom_grammar s =
+ match find_grammars_by_name s with
+ | [AnyEntry c;AnyEntry p] -> (Obj.magic c, Obj.magic p)
+ | _ -> assert false
+
```

Comment:
This is a footgun I don't remember seeing on the first iteration of the patch, but my memory may be failing.

**DON'T USE OBJ.MAGIC FOR THIS KIND OF CODE.**

You don't have to create new generic arguments either, this is probably a remnant of the time this was working with tactic notations. Instead of doing so, use the fact that you know statically these entries are going to be of type `constr` and `constr_pattern` and use a dynamic cast on the dynamic witness type.